src/layers/core.ts

```diff
@@ -607,3 +607,94 @@ export class Reshape extends Layer {
   }
 }
 serialization.SerializationMap.register(Reshape);
+
+export interface PermuteLayerConfig extends LayerConfig {
+  /**
+   * Array of integers. Permutation pattern. Does not include the
+   * sample (batch) dimension. Index starts at 1.
+   * For instance, `[2, 1]` permutes the first and second dimensions
+   * of the input.
+   */
+  dims: number[];
+}
+
+/**
+ * Permutes the dimensions of the input according to a given pattern.
+ *
+ * Useful for, e.g., connecting RNNs and convnets together.
+ *
+ * Example:
+ *
+ * ```js
+ * const model = tf.Sequential();
+ * model.add(tf.layers.permute({
+ *   dims: [2, 1],
+ *   inputShape: [10, 64]
+ * }));
+ * console.log(model.outputShape);
+ * // Now model's output shape is [null, 64, 10], where null is the
+ * // unpermuted sample (batch) dimension.
+ * ```
+ *
+ * Input shape:
+ *   Arbitrary. Use the configuration field `inputShape` when using this
+ *   layer as othe first layer in a model.
+ *
+ * Output shape:
+ *   Same rank as the input shape, but with the dimensions re-ordered (i.e.,
+ *   permuted) according to the `dims` configuration of this layer.
+ */
+export class Permute extends Layer {
+  static className = 'Permute';
+  readonly dims: number[];
+  private readonly dimsIncludingBatch: number[];
+
+  constructor(config: PermuteLayerConfig) {
+    super(config);
+    if (config.dims == null) {
+      throw new Error(
+          'Required configuration field `dims` is missing during Permute ' +
+          'constructor call.');
+    }
+    if (!Array.isArray(config.dims)) {
+      throw new Error(
+          'Permute constructor requires `dims` to be an Array, but received ' +
+          `${config.dims} instead.`);
+    }
+
+    // Check the validity of the permutation indices.
+    const expectedSortedIndices = range(1, config.dims.length + 1);
+    if (!util.arraysEqual(config.dims.slice().sort(), expectedSortedIndices)) {
+      throw new Error(
+          'Invalid permutation `dims`: ' + JSON.stringify(config.dims) +
+          ' `dims` must contain consecutive integers starting from 1.');
```

Comment:
Nice error string :)
Maybe add the expected last index as well?  or how many were expected (i guess that's the same number anyway).

Reply:
I was surprised that the batch dim is always skipped. may want to add something to that effect as well.  "(The batch dim is always unchanged by 'Permute')"